app/code/Magento/CatalogUrlRewrite/Observer/ProductProcessUrlRewriteSavingObserver.php

```diff
@@ -32,30 +43,62 @@ class ProductProcessUrlRewriteSavingObserver implements ObserverInterface
      */
     private $productUrlPathGenerator;
 
+    /**
+     * @var StoreManagerInterface
+     */
+    private $storeManager;
+
+    /**
+     * @var ProductScopeRewriteGenerator
+     */
+    private $productScopeRewriteGenerator;
+
+    /**
+     * @var DeleteEntitiesFromStores
+     */
+    private $deleteEntitiesFromStores;
+
+    /**
+     * @var Collection
+     */
+    private $productCollection;
+
     /**
      * @param ProductUrlRewriteGenerator $productUrlRewriteGenerator
      * @param UrlPersistInterface $urlPersist
-     * @param ProductUrlPathGenerator|null $productUrlPathGenerator
+     * @param ProductUrlPathGenerator $productUrlPathGenerator
+     * @param StoreManagerInterface $storeManager
+     * @param ProductScopeRewriteGenerator $productScopeRewriteGenerator
+     * @param DeleteEntitiesFromStores $deleteEntitiesFromStores
+     * @param Collection $productCollection
      */
     public function __construct(
         ProductUrlRewriteGenerator $productUrlRewriteGenerator,
         UrlPersistInterface $urlPersist,
-        ProductUrlPathGenerator $productUrlPathGenerator = null
+        ProductUrlPathGenerator $productUrlPathGenerator,
+        StoreManagerInterface $storeManager,
+        ProductScopeRewriteGenerator $productScopeRewriteGenerator,
+        DeleteEntitiesFromStores $deleteEntitiesFromStores,
+        Collection $productCollection
```

Comment:
Product collection contains state, so for collection you need to use factory in constructor and create collection object only when needed. I think this might cause test failures

Reply:
Yes, I'm learning. That explains why when repeatedly getting product data using product collection, it was not showing updates that were being saved with repository. Thanks. I have updated the code.